crates/update-notifier/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+use semver::{Error as SemVerError, Version};
+use serde::{Deserialize, Serialize};
+use serde_json::Error as SerdeError;
+use thiserror::Error as ThisError;
+use tiny_gradient::{GradientStr, RGB};
+
+mod fetch;
+mod ui;
+mod utils;
+
+// default interval to check for new updates (one per day)
+const DEFAULT_INTERVAL: Duration = Duration::from_secs(60 * 60 * 24);
+
+#[derive(Debug)]
+pub struct UpdateNotifier {
+    config: UpdateNotifierConfig,
+    package: String,
+    tag: Option<String>,
+    interval: Duration,
+}
+
+// This portion of the config is serialized and persisted to disk
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct UpdateNotifierConfig {
+    current_version: String,
+    latest_version: Option<String>,
+    last_checked: Option<u128>,
+}
+
+#[derive(ThisError, Debug)]
+pub enum UpdateNotifierError {
+    #[error("Failed to fetch latest version")]
+    FetchError(#[from] ReqwestError),
+    #[error("Failed to parse version")]
+    ParseError(#[from] SemVerError),
+    #[error("Failed to parse JSON")]
+    JsonError(#[from] SerdeError),
+    #[error("Failed to write to terminal")]
+    DisplayLengthError(#[from] ui::utils::GetDisplayLengthError),
+    #[error("Unknown Error")]
+    Internal(),
+}
+
+impl UpdateNotifier {
+    fn should_refresh(&self) -> bool {
+        if self.config.latest_version.is_none() {
+            log::debug!("no latest version found in local config");
+            return true;
+        }
+        match self.config.last_checked {
+            Some(last_checked) => {
+                let now = utils::ms_since_epoch();
+                let diff = now - last_checked;
+                log::debug!(
+                    "last checked {} ms ago (interval={} ms)",
+                    diff,
+                    self.interval.as_millis()
+                );
+                log::debug!("refreshing in {} ms", self.interval.as_millis() - diff);
+                diff > self.interval.as_millis()
+            }
+            None => {
+                log::debug!("no last checked time found in local config");
+                true
+            }
+        }
+    }
+
+    fn update_message(&self) -> Result<(), UpdateNotifierError> {
+        let turbo = "@turborepo";
+        let turbo_gradient = turbo.gradient([RGB::new(0, 153, 247), RGB::new(241, 23, 18)]);
+
+        let latest_version = match &self.config.latest_version {
+            Some(v) => v,
+            None => {
+                log::error!("no latest version found in local config");
+                return Err(UpdateNotifierError::Internal());
+            }
+        };
+
+        let msg = format!(
+            "
+            Update available {} ≫ {}
+            Changelog: {}/releases/tag/v{}
+            Run \"{}\" to update
+
+            Follow {} for updates: {}
+            ",
+            &self.config.current_version.dimmed(),
+            &latest_version.green().bold(),
+            "https://github.com/vercel/turbo",
```

Comment:
I'd like to see this "cleaned up" a bit more

ideally have 0 references to anything turbo in this crate and pass everything in

right now the constructor gets the package and version to check, but we have other links hardcoded here

Reply:
Yea, this is something that bothers me too, I started down the road of making this fully generic but didn't completely finish it up (was punting until we shipped 1.7)

Reply:
Yeah I think that's something worth making a ticket and revisiting